verify.js

```diff
@@ -86,96 +72,131 @@ module.exports = function (jwtString, secretOrPublicKey, options, callback) {
-
-  if (options.issuer) {
-    var invalid_issuer =
-        (typeof options.issuer === 'string' && payload.iss !== options.issuer) ||
-        (Array.isArray(options.issuer) && options.issuer.indexOf(payload.iss) === -1);
-
-    if (invalid_issuer) {
-      return done(new JsonWebTokenError('jwt issuer invalid. expected: ' + options.issuer));
-    }
-  }
-
-  if (options.subject) {
-    if (payload.sub !== options.subject) {
-      return done(new JsonWebTokenError('jwt subject invalid. expected: ' + options.subject));
-    }
-  }
-
-  if (options.jwtid) {
-    if (payload.jti !== options.jwtid) {
-      return done(new JsonWebTokenError('jwt jwtid invalid. expected: ' + options.jwtid));
-    }
-  }
-
-  if (options.maxAge) {
-    if (typeof payload.iat !== 'number') {
-      return done(new JsonWebTokenError('iat required when maxAge is specified'));
-    }
-
-    var maxAgeTimestamp = timespan(options.maxAge, payload.iat);
-    if (typeof maxAgeTimestamp === 'undefined') {
-      return done(new JsonWebTokenError('"maxAge" should be a number of seconds or string representing a timespan eg: "1d", "20h", 60'));
-    }
-    if (clockTimestamp >= maxAgeTimestamp + (options.clockTolerance || 0)) {
-      return done(new TokenExpiredError('maxAge exceeded', new Date(maxAgeTimestamp * 1000)));
-    }
-  }
-
-  return done(null, payload);
+  var getSecret;
+
+  if(typeof secretOrPublicKey === 'function') {
+      if(!callback) {
+          return done(new JsonWebTokenError('verify must be called asynchronous if secret or public key is provided as a callback'));
+      }
+
+      getSecret = secretOrPublicKey;
+  }
+  else {
+      getSecret = function(header, secretCallback) {
+          return secretCallback(null, secretOrPublicKey);
+      };
+  }
+
+  return getSecret(header, function(err, secretOrPublicKey) {
+      if(err) {
+          return done(new JsonWebTokenError('error in secret or public key callback: ' + err.message));
+      }
+
+      if (!hasSignature && !options.algorithms) {
+          options.algorithms = ['none'];
+      }
+
+      if (!options.algorithms) {
+          options.algorithms = ~secretOrPublicKey.toString().indexOf('BEGIN CERTIFICATE') ||
+          ~secretOrPublicKey.toString().indexOf('BEGIN PUBLIC KEY') ?
+              ['RS256', 'RS384', 'RS512', 'ES256', 'ES384', 'ES512'] :
+              ~secretOrPublicKey.toString().indexOf('BEGIN RSA PUBLIC KEY') ?
+                  ['RS256', 'RS384', 'RS512'] :
+                  ['HS256', 'HS384', 'HS512'];
+
+      }
+
+      if (!~options.algorithms.indexOf(header.alg)) {
+          return done(new JsonWebTokenError('invalid algorithm'));
+      }
+
+      var valid;
+
+      try {
+          valid = jws.verify(jwtString, header.alg, secretOrPublicKey);
+      } catch (e) {
+          return done(e);
+      }
+
+      if (!valid)
+          return done(new JsonWebTokenError('invalid signature'));
+
+      var payload;
+
+      try {
+          payload = decode(jwtString);
```

Comment:
🤔 wasn't this the `decode` you wanted to clean in favor of the `decode(..., { complete: true })` call you add at the beginning of the `verify()` function?

Reply:
FYI Not sure why but in github if I see the diff omitting whitespaces it shows me you removed this, however, in the normal view or even if I click in View button for this file I can still see this decode call

Reply:
You are right, should be okay now :)